homeassistant/components/sensor/history_values.py

```diff
@@ -0,0 +1,369 @@
+        now_timestamp = math.floor(dt_util.as_timestamp(now))
+
+        # If period has not changed and current time after the period end...
+        if (start_timestamp == p_start_timestamp
+                and end_timestamp == p_end_timestamp
+                and end_timestamp <= now_timestamp):
+            # Don't compute anything as the value cannot have changed
+            _LOGGER.debug("Update not necessary.")
+            return
+
+        # Get history between start and end
+        history_states = await self.hass.async_add_job(
+            history.state_changes_during_period, self.hass, start, end,
+            str(self._target_entity))
+        history_list = history_states.get(self._target_entity)
+        if history_list is None:
+            self.value = STATE_UNKNOWN
+            return
+
+        if self._type in [TYPE_LOW, TYPE_PEAK, TYPE_RANGE]:
+            value = self.get_min_max_range(history_list)[self._type]
+        if self._type == TYPE_AVG:
+            value = self.get_average(history_list, end, now)
+        if value is not None:
+            self.value = value
+
+        self.async_schedule_update_ha_state()
+
+    @property
+    def device_class(self):
+        """Return the device_class of the sensor."""
+        return self._device_class
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._friendly_name
+
+    @property
+    def should_poll(self):
+        """Return False, we schedule our own updates."""
+        return False
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self.value
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit_of_measurement of the sensor."""
+        return self._unit_of_measurement
+
+    @callback
+    def get_average(self, state_list, end, now):
+        """Calculate and return the average value of @state_list."""
+        if state_list is None:
+            return
+        total_value = 0
+        precision = 0
+        total_period = datetime.timedelta(seconds=0)
+        last_idx = len(state_list)-1
+        for state in state_list:
+            if state.state == STATE_UNKNOWN:
+                # Don't take this into account at all.
+                continue
+            current_prec = self.get_precision(float(state.state))
+            if current_prec > precision:
+                precision = current_prec
+            next_updated = state_list.index(state) + 1
+            if next_updated > last_idx:
+                if end > now:
+                    next_time = now
+                else:
+                    next_time = end
+            else:
+                next_time = state_list[next_updated].last_changed
+            state_period = next_time - state.last_changed
+            total_period += state_period
+            total_value += float(state.state) * state_period.total_seconds()
```

Comment:
Looking into this now, but there does not seem to be a way to reduce() both total_period and total_value without calculating state_period multiple times for each state (once per reduce() call) or storing it somewhere for later use. Am I missing something?

Reply:
I've been thinking about this sensor... why don't you go directly to the RDBMS to get these values ? You require the aggregate values, but transfer lots of information out of the RDBMS...

Could an SQL query get the information you need ?

Reply:
Let me start out by saying that I don't know exactly how the data is stored in the database. From my experience with SQL so far I can imagine that the `state_period`s would still need to be calculated and added up outside of the DB engine. `total_value` is a sum of the product of each state and its duration and since the duration would probably need to be calculated outside of the DB engine, so does this.
Apart from the issues above, I personally would not want to maintain a project which has DB access and SQL code scattered all over the place. I would prefer to keep all of that stuff in a separate module to make changing/upgrading the database schema in the future easier.

Reply:
Looked into reduce(), found no way to make significant improvements.